### DISCUSSION TITLE
MAC interrupts not firing

### DISCUSSION THREAD

**Discussion by mcergun** [polarfire-soc/polarfire-soc-documentation#151] [How to] [answered]
Hi. I'm trying to get ethernet up and running on reference icicle board.
However I can't seem to be able to successfully send or receive packets. First call to `_send_pkt` returns `MSS_MAC_ERR_OK` but the next 4 of the subsequent calls return `MSS_MAC_ERR_NOT_DONE` and no interrupts are fired on the device. I can see the sent packets on the wire via WireShark on my desktop computer (which is directly connected to the icicle kit)

The code is copied from eMMC to DDR and it runs on u54_1 only.

My initialize routine looks like this:
```c
// GPIO initialization
// Following are taken from example projects for GEM0_LOCAL configuration
MSS_GPIO_init(GPIO2_LO);

MSS_GPIO_config(GPIO2_LO, MSS_GPIO_26, MSS_GPIO_OUTPUT_MODE);
MSS_GPIO_config(GPIO2_LO, MSS_GPIO_27, MSS_GPIO_OUTPUT_MODE);
MSS_GPIO_config(GPIO2_LO, MSS_GPIO_28, MSS_GPIO_OUTPUT_MODE);

MSS_GPIO_config(GPIO2_LO, MSS_GPIO_30, MSS_GPIO_INPUT_MODE);
MSS_GPIO_config(GPIO2_LO, MSS_GPIO_31, MSS_GPIO_INPUT_MODE);

MSS_GPIO_set_output(GPIO2_LO, MSS_GPIO_26, 0);
MSS_GPIO_set_output(GPIO2_LO, MSS_GPIO_27, 0);
MSS_GPIO_set_output(GPIO2_LO, MSS_GPIO_28, 0);
```

```c
// MAC initialization
MSS_MAC_cfg_struct_def_init(&cfg);

// Set MAC address here
cfg.speed_duplex_select = MSS_MAC_ANEG_ALL_SPEEDS;
cfg.phy_addr = PHY_VSC8662_1_MDIO_ADDR;
cfg.phy_type = MSS_MAC_DEV_PHY_VSC8662;
cfg.pcs_phy_addr = SGMII_MDIO_ADDR;
cfg.interface_type = TBI;
cfg.phy_mac_autonegotiate = MSS_MAC_VSC8662_mac_autonegotiate;
// Function pointer assignemnts for
// phy_get_link_status, phy_init, phy_set_link_speed functions
cfg.use_local_ints = MSS_MAC_ENABLE;

cfg.mac_addr[5]++;
MSS_MAC_init(&g_mac1, &cfg);

cfg.mac_addr[5]--;
cfg.phy_addr = PHY_VSC8662_0_MDIO_ADDR;
cfg.phy_controller = &g_mac1;
MSS_MAC_init(&g_mac0, &cfg);

MSS_MAC_set_tx_callback(&g_mac0, 0, tx_callback);
MSS_MAC_set_rx_callback(&g_mac0, 0, rx_callback);

MSS_MAC_receive_pkt(&g_mac0, 0, buf[0], 0, MSS_MAC_INT_DISABLE);
MSS_MAC_receive_pkt(&g_mac0, 0, buf[1], 0, MSS_MAC_INT_DISABLE);
MSS_MAC_receive_pkt(&g_mac0, 0, buf[2], 0, MSS_MAC_INT_DISABLE);
MSS_MAC_receive_pkt(&g_mac0, 0, buf[3], 0, MSS_MAC_INT_ARM);
```

However, after that point I can't get any interrupts for neither tx or rx events. Is there a point I'm missing?

EDIT: I can receive other local interrupts such as MMUART on that hart. Also I manually checked the MIE and MSTATUS registers, their values seem valid.

```
MSTATUS: 8000 000A 0000 6088
.MIE: 1

MIE: 0900 0880 (taken after MSS_MAC_send_pkt call)
.MTIE: 1
.MEIE: 1
.LI8:  1 (mac0_int on U54_1 and U54_2)
.LI11: 1 (mmuart2 on U54_1)
```

**Comment by AntonKrug** [comment]
This might be completely irelevant to your case and I do not have enought knowledge about this topic to know for sure. But once I had a networking issue where only one in handful of packets was working. In the end the peripheral and my buffer in the stack had different aligments and only one in few packets was aligned correctly and working, while the rest of them was confusing me a lot. Was looking for good while at wireshark to get to the bottom of my problem. So I'm just curious could you have some misconfiguration between the peripheral and your software stack as well?

**Reply by mcergun** [reply]
I don't think there's an issue with the buffer alignments. I created them on the stack with `__attribute((aligned(4))`
Also I can see many ARP packets being formed correctly if I use the same initialization and use LWIP & FreeRTOS. The problem is that the eth stack is not aware (or can't communicate with the code) that the packets are sent or they have arrived.

**Comment by hughbreslin** [comment]
Hey @mcergun 

could you try using PLIC interrupts instead of local and see if it works?


    cfg.use_local_ints = MSS_MAC_ENABLE;

to

    cfg.use_local_ints = MSS_MAC_DISABLE;

**Reply by mcergun** [reply]
I've added `PLIC_init` to initialization and and set `use_local_ints` as `MSS_MAC_DISABLE`. Now the interrupts are firing and my callback functions are called.
I wanted to use local interrupts as they are easier to integrate with FreeRTOS (I've just updated interrupt vector table with my custom function, no other change was needed)

Do you have any clues why they are not working on local interrupt mode?

**Reply by hughbreslin** [reply]
Cool :) 

Check out the answer from Jim to [post 142](https://github.com/orgs/polarfire-soc/discussions/142#discussioncomment-3264592) it might give you some insight as I think ye ran into the same issue

**Reply by mcergun** [reply]
He was getting interrupts once or twice but no more times. I think my problem is a little bit different; I can never receive the interrupts.

Also, I've tried local interrupts with and without the FreeRTOS. So in the bare-metal case, there should be no "pre-emption" as he stated. The interrupt handlers are set once, by the MPFS HAL.

**Reply by hughbreslin** [reply]
On the PLIC can you see what the interrupt numbers are that are coming in? I.e are they 51:56 or are they 57:62? You should also be able to see MIP using the local interrupts and which ones are pending and if they actually ever are. The local interrupt mapping is a bit different so thats why I'm wondering

**Reply by mcergun** [reply]
> On the PLIC can you see what the interrupt numbers are that are coming in? I.e are they 51:56 or are they 57:62? You should also be able to see MIP using the local interrupts and which ones are pending and if they actually ever are. The local interrupt mapping is a bit different so thats why I'm wondering

`int_num` returned by `PLIC_ClaimIRQ()` is 70, so it's not in the range for neither of them.

For the MIP register, shouldn't I receive an interrupt if there's a pending one? I will still check it's value but that doesn't make sense for me.

**Reply by hughbreslin** [reply]
The PLIC interrupts are offset by 13 so this is actually interrupt 57 which is the mac1_int (from the TRM section 6):

![image](https://user-images.githubusercontent.com/44862881/181733143-665e346a-ebb9-4cc1-afc5-e3c5a3691640.png)

This is just a suspicion as its not something I've directly worked on, I'll check and hopefully know definitively on Tuesday - as MAC1 is being used for PHY control its actually generating the interrupts. The local IRQs for MAC1 are only available on U54_3 and U54_4 as you can see in the table above. Which means if this is actually where the interrupts are coming from you'll never see them on U54_1 in a local interrupt set up. Your non-FreeRTOS set up will probably run fine on U54_3 or U54_4 with the local IRQs. 

The reason I wanted to check MIP is for this exact reason, you'd probably see the MIP bit set for the direct interrupt on U54_3 and U54_4 from MAC1 when you're expecting it on U54_1.

**Reply by mcergun** [reply]
Hey @hughbreslin quick update: I must have left `_send_pkt` function with `g_mac1` between my tries. When I set it back to `g_mac0` as it was supposed to be, the IRQs now arrive with 64 (51 for global). So sorry for the mistake.

This means the interrupt is generated by the correct GEM, right? So there must be different reason for the problem as far as I understand.

**Reply by hughbreslin** [reply]
Hey, no problem, that definitely wouldn't help anyway :) 

> This means the interrupt is generated by the correct GEM, right?

Yes 64 sounds a bit nicer, you should be able to see them now in the bare metal set up as local ints on U54_1. You could try setting a breakpoint on the `handle_local_interrupt` call in [`mss_mtrap.c`](https://github.com/polarfire-soc/platform/blob/main/mpfs_hal/common/mss_mtrap.c#L735) and if the interrupt is definitely triggering and enabled you'll hit the breakpoint (maybe the handler isn't being linked correctly or something along those lines). If you don't hit the breakpoint the interrupt is never being taken but you should see it pending at least if its triggering - I assume its at least triggering as you're seeing it come in on the PLIC.

**Reply by mcergun** [reply]
Sadly the local interrupts still don't arrive.

I've checked following fields:

- MSTATUS.MIE[3] = 1
- MIE.LI8[24] = 1
- MIP = always 0x0

My breakpoint is generally on `trap_from_machine_mode` to be able to track any interrupts, which there is none.

**Reply by hughbreslin** [reply]
Aha I think I found it - theres one last step... well its actually a first step, are you setting the FAB_INTEN_MISC register before enabling interrupts? This register is used to select between local and direct interrupts for the MAC:

![image](https://user-images.githubusercontent.com/44862881/181765420-95dde1b8-b639-4241-a1db-9fe1a06dc2ea.png)

Doing the following write should route the interrupts correctly:

    SYSREG->FAB_INTEN_MISC |= FAB_INTEN_MISC_MAC0_U54_1_MASK;

**Reply by mcergun** [reply]
Wow this seems to resolve the issue :).

Previously I tried setting all bits in `FAB_INTEN_U54_1` as stated in the TechRefManual below Table 6-1 but didn't know about the `FAB_INTEN_MISC` register.

Can't accept your message as an answer probably because it is a reply to a comment. I can mark it as an answer if you post the same information in another comment.

**Reply by hughbreslin** [reply]
Woohoo! :) I'm glad thats got it going for you! 

Posted a new comment so it shows up at the top.

**Comment by hughbreslin** [comment, marked as answer]
Posting from the main thread so the answer is easy to find:

Aha I think I found it - theres one last step... well its actually a first step, are you setting the FAB_INTEN_MISC register before enabling interrupts? This register is used to select between local and direct interrupts for the MAC:

![image](https://user-images.githubusercontent.com/44862881/181765420-95dde1b8-b639-4241-a1db-9fe1a06dc2ea.png)

Doing the following write should route the interrupts correctly:

    SYSREG->FAB_INTEN_MISC |= FAB_INTEN_MISC_MAC0_U54_1_MASK;
